Single step oro & fix files generation procedure for atm and ocn res combo

sanatcumar opened the pull request:
## DESCRIPTION OF CHANGES: 
Consolidated the multi-step process for generating the oro and fix files for combinations of atmospheric and ocean grid resolutions. 

## TESTS CONDUCTED: 
- [x] Compile branch on all Tier 1 machines using Intel (Orion, Jet, Hera and WCOSS2). Done using d0d1e8f.
- [x] Compile branch on Hera using GNU.  Done using d0d1e8f.
- [x] Compile branch in 'Debug' mode on WCOSS2. Done using d0d1e8f on Dogwood.
- [x] Run unit tests locally on any Tier 1 machine. Done on Hera using d0d1e8f. All tests passed.
- [x] Run relevant consistency tests locally on all Tier 1 machine. Done using d0d1e8f. All tests other than `grid_gen` tests passed. This result was expected. For details on the `grid_gen` tests, see below.

Describe any additional tests performed.
 Pasted below is a summary of the tests [conducted.](https://docs.google.com/document/d/1qQvZfPxpw6M2qnzLbatOX5kRsD9a0nGcJjdAYDJogWk/edit)

New scripts and codes were written to reduce the multi step procedure of generating coupled grids into a single step process. 

During the development of this new single step process, the original “orog” code was split into two parts that is sequentially activated by flags in the scripts to first generate a land mask and then use the land mask generated by the ocean merge procedure to populate the orography fields. The following information documents the effects of these code changes.

Five comparisons with baselines were made. 

1. Same grids, same data sets and different orography code
> Ocean merged files **(Identical)**
> Orography **(Minor differences along spatial pattern at zero lat, 0, 90, 180 & 270 lon)**
> Vegetation type **(Mostly identical)**
 

2. Different grids, same data sets, different orography code (Comparison with /scratch1/NCEPDEV/global/glopara/fix/orog/20230615)
> Grids  **(Different but minor )**
> Orography **(Minor differences along spatial pattern at zero lat, 0, 90, 180 & 270 lon)**
> Vegetation type **(Mostly identical)** 

3. Same grids, different data sets, vegfrac=False, different orography code 
> Vegetation type **(Mostly identical)**

4. Same grids, different data sets, vegfrac=True, different orography code 
> Vegetation type **(Mostly identical)**

5. Comparisons of orography files and select sfc fields for grid types other than uniform with files created using the latest develop branch
> Orography **(Identical)**
> Vegetation type **(Identical)**

## DEPENDENCIES:
#843 must be merged first.  https://github.com/ufs-community/UFS_UTILS/issues/854 needs to be merged first

## DOCUMENTATION:
The repository was compiled on Hera with Doxygen (d0d1e8f). The `orog.fd` and `ocean_merge.fd` had no Doxygen warnings.

## ISSUE: 
Fixes #825.

## CONTRIBUTORS: 
@GeorgeGayno-NOAA , @shansun6 , @aerorahul , @barlage

GeorgeGayno-NOAA committed:
Add logic to ignore soil and clay percentage records.

Part of #821.

GeorgeGayno-NOAA committed:
Merge branch 'develop' into bugfix/bnu_soil

Part of #821.

GeorgeGayno-NOAA committed:
Update documentation and script logs for new files.

Part of #821.

sanatcumar committed:
updates done by Sanatcumar

sanatcumar committed:
updates done by Sanatcumar

sanatcumar committed:
updates done by Sanatcumar

sanatcumar committed:
updates done by Sanatcumar

sanatcumar committed:
updates done by Sanatcumar

sanatcumar committed:
updates done by Sanatcumar

sanatcumar committed:
updates done by Sanatcumar

sanatcumar committed:
Removed emails and links

sanatcumar commented:
Some notes.

Comparisons of results ( including outputs of individual steps ) showed some differences and are listed below
- In Grid generation

- - Grids generated using the latest codes are different from the ./fix/orog for example
nccmp -dsSfq  /scratch2/NCEPDEV/land/Sanath.Kumar/UFS_UTILS_SK/fix/orog/C96/C96_grid.tile1.nc /scratch2/NCEPDEV/stmp1/Sanath.Kumar/my_grids_single_step/C96/C96_grid.tile1.nc is **NOT IDENTICAL**

Variable Group Count          Sum      AbsSum          Min         Max       Range         Mean      StdDev
x        /      9433  2.72155e-11 2.98002e-10 -5.68434e-14 6.03961e-14  1.1724e-13  2.88513e-15  3.7784e-14
y        /     20841  1.52045e-13 9.86002e-11 -2.84217e-14 2.84217e-14 5.68434e-14  7.29548e-18 6.11732e-15
dx       /     37056     -56031.9     56031.9     -1.81328    -1.28225    0.531026     -1.51209    0.144961
dy       /     37056     -56031.9     56031.9     -1.81328    -1.28225    0.531026     -1.51209    0.144961
area     /     36864 -5.33744e+09 5.33744e+09      -209474    -91134.2      118340      -144787     30233.2
angle_dx /     28925  -6.6303e-13  8.7861e-09 -3.59535e-12 3.59535e-12 7.19069e-12 -2.29224e-17 4.18713e-13
angle_dy /      7011 -7.81597e-13 8.55029e-09 -7.83018e-12 7.81597e-12 1.56462e-11 -1.11482e-16 1.91881e-12

**Differences are small** but different runs are consistent. For example

nccmp -dsSfq  /scratch2/NCEPDEV/stmp1/Sanath.Kumar/my_grids_multi_step/C96/C96_grid.tile1.nc /scratch2/NCEPDEV/stmp1/Sanath.Kumar/my_grids_single_step/C96/C96_grid.tile1.nc **is identical**


- Step 3 Oro generation using UFS UTILS current codes and older codes in Shan's directory are different
For example 
/scratch2/NCEPDEV/stmp1/Sanath.Kumar/my_grids_multi_step/step3/C96.mx500_oro_data.tile1.nc /scratch2/NCEPDEV/stmp1/Sanath.Kumar/my_grids_multi_step/step3_mod/C96.mx500_oro_data.tile1.nc
is not identical 
Step3 contains the outputs using Shan's executable "orog"
Step3_mod contains the output using the latest executable "orog".

All other files used in this process are identical. Source of differences is unresolved.

-  final comparison examples

nccmp -dsSfq /scratch2/NCEPDEV/stmp1/Sanath.Kumar/my_grids_multi_step/step3_mod/C96.mx500_oro_data.tile1.nc /scratch2/NCEPDEV/stmp1/Sanath.Kumar/my_grids_single_step/C96.mx500/C96_oro_data.tile1.nc is different only in the orog_filt fields. This is due to minor file name differences in Shan's code. 

nccmp -dsSfq /scratch2/NCEPDEV/stmp1/Sanath.Kumar/my_grids_single_step/C96.mx500/fix_sfc/C96.vegetation_type.tile1.nc /scratch2/NCEPDEV/stmp1/Sanath.Kumar/my_grids_multi_step/sfc.C96.mx500_mod/C96.mx500.vegetation_type.tile1.nc is identical

GeorgeGayno-NOAA commented on input.nml:
Please remove.

sanatcumar replied:
done

GeorgeGayno-NOAA commented on ush/fort.15:
Please remove.

sanatcumar replied:
done

GeorgeGayno-NOAA commented on ush/fort.235:
Please remove.

sanatcumar replied:
done

GeorgeGayno-NOAA commented on ush/fort.25:
Please remove.

sanatcumar replied:
done

GeorgeGayno-NOAA commented on ush/landcover30.fixed:
Please remove.

sanatcumar replied:
Done

sanatcumar committed:
Removed input files

sanatcumar committed:
Removed input files

sanatcumar committed:
Removed log files

GeorgeGayno-NOAA commented on ush/fv3gfs_driver_grid.sh:
```diff
@@ -468,6 +486,20 @@ elif [ $gtype = regional_gfdl ] || [ $gtype = regional_esg ]; then
+#------------------------------------------------------------------------------------
+#------------------------------------------------------------------------------------
+#Redo orog and filter.
+#------------------------------------------------------------------------------------
```
Why is the orography re-created?

sanatcumar replied:
As per notes , the lake mask is derived first using the step 1 and when the ocean mask is merged, the lake/land points may change and hence is re run.  The lake merge routine needs the oro files to derive the initial lakemask and hence is re run after it is modified.

GeorgeGayno-NOAA replied:
Ok. I would add a few comments to the script for clarity. Also, I would document this multi-step process in the issue (#825).

sanatcumar replied:
I have added some explanatory texts to both scripts and the issue

GeorgeGayno-NOAA commented:
I tried running 0b9a03e on Hera. I used the default settings. I got this error:

```
+ /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../ush/fv3gfs_ocean_merge.sh 96 500
ocean_merge  exe file is found in  /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../exec/
 flnm_ocn=
 /scratch1/NCEPDEV/stmp4/Sanath.Kumar/CPLD_GRIDGEN//500/C96.mx500.tile1.nc
                                           ncid=       65536  ndims=           2
 nvars=           3  natts=           0
 lat=          96 lon=          96
  flnm2=/scratch2/NCEPDEV/stmp1/George.Gayno/my_grids/C96/C96_oro_data.tile1.nc
 No such file or directory

999
error in fv3gfs_ocean_merge.sh
```

sanatcumar committed:
fixed the typo in ocean merge paths

sanatcumar commented:
Thanks for catching this bug.  I just fixed and tested it . The issue  was incorrect paths in the ocean merge routine.  I just fixed and tested the default settings with empty directories.

sanatcumar committed:
added explanatory comments as recomended to the ocean_merge routines

GeorgeGayno-NOAA commented on sorc/ocean_merge.fd/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+target_link_libraries(
+  ${exe_name}
+  sp::sp_d
+  NetCDF::NetCDF_Fortran)
```
The NCEPLIBS group is trying to get rid of the sp library. I don't see where sp is used by this code.

sanatcumar replied:
I did not check for redundant libraries.  I just removed that and was able to compile successfully.

sanatcumar committed:
removed unnecessary libraries (sp) when compiling

GeorgeGayno-NOAA commented on ush/fv3gfs_driver_grid.sh:
```diff
@@ -479,11 +515,31 @@ fi
+module use ${BASE_DIR}/modulefiles
+module load build.hera.intel
+module list
+
```
Do these modules need to be loaded again in this script?

sanatcumar replied:
Hi George, I tried commenting these and I got this error ""error while loading shared libraries: libiomp5.so: cannot open shared object file:"
I will remove the module list for the time being

GeorgeGayno-NOAA commented:
The branch at 069d1bf is breaking during the second filtering step. I am using the default values to set up the grid:

```
 - N_DEL2_WEAK:            8
 - MAX_SLOPE:   0.120000000000000
 - PEAK_FAC:    1.10000000000000
 Read the grid from file /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTI
 LS.sanath/driver_scripts/../fix/orog/C96/C96_mosaic.nc
  read_grid_file: This is a global grid.
 Before filter: Max_slope=  3.320347328976547E-002
 After filter: Max_slope=  3.320347328976547E-002

real    0m1.936s
user    0m1.878s
sys     0m0.026s
cp: cannot stat '/scratch2/NCEPDEV/stmp1/George.Gayno/my_grids/C96/C96_mosaic.nc': No such file or directory
```

sanatcumar committed:
fixed a broken path in ocean_merge

sanatcumar commented:
> The branch at [069d1bf](https://github.com/ufs-community/UFS_UTILS/commit/069d1bf611ef2d10eceae4fd7136dfb2c6707df4) is breaking during the second filtering step. I am using the default values to set up the grid:
> 
> sys     0m0.026s
> cp: cannot stat '/scratch2/NCEPDEV/stmp1/George.Gayno/my_grids/C96/C96_mosaic.nc': No such file or directory
> ```


 This was related to the first bug in the file paths.  It should be fixed in this update

GeorgeGayno-NOAA commented on ush/orog.sh:
```diff
@@ -0,0 +1,112 @@
+  echo error
+  exit $iret
+fi
+
```
I had to adjust this line as follows to get it to work.

```
-cp $out_dir/C${CRES}/C${CRES}_mosaic.nc $WORKDIR
+cp $out_dir/C${CRES}_mosaic.nc $WORKDIR
```

sanatcumar replied:
You are right and I am puzzled on how it ran for me.  The out_dir as exported originally is just the base directory but then the CXX is concatenated in the later scripts.  I have edited it in the latest commit

GeorgeGayno-NOAA commented:
I just tried 41f87d6 for a C96 grid. I got two sets of orog files that are different sizes. What are these files?

```
/scratch2/NCEPDEV/stmp1/George.Gayno/my_grids_single_step/C96.mx500 $ ls -l *oro*
-rw-r--r-- 1 George.Gayno stmp 757634 Jul  5 20:09 C96_oro_data.tile1.nc
-rw-r--r-- 1 George.Gayno stmp 757634 Jul  5 20:09 C96_oro_data.tile2.nc
-rw-r--r-- 1 George.Gayno stmp 757634 Jul  5 20:09 C96_oro_data.tile3.nc
-rw-r--r-- 1 George.Gayno stmp 757634 Jul  5 20:09 C96_oro_data.tile4.nc
-rw-r--r-- 1 George.Gayno stmp 757634 Jul  5 20:09 C96_oro_data.tile5.nc
-rw-r--r-- 1 George.Gayno stmp 757634 Jul  5 20:09 C96_oro_data.tile6.nc
```

```
/scratch2/NCEPDEV/stmp1/George.Gayno/my_grids_single_step/C96 $ ls -l *oro*
-rw-r--r-- 1 George.Gayno stmp 882562 Jul  5 20:07 C96_oro_data.tile1.nc
-rw-r--r-- 1 George.Gayno stmp 882562 Jul  5 20:07 C96_oro_data.tile2.nc
-rw-r--r-- 1 George.Gayno stmp 882562 Jul  5 20:07 C96_oro_data.tile3.nc
-rw-r--r-- 1 George.Gayno stmp 882562 Jul  5 20:07 C96_oro_data.tile4.nc
-rw-r--r-- 1 George.Gayno stmp 882562 Jul  5 20:07 C96_oro_data.tile5.nc
-rw-r--r-- 1 George.Gayno stmp 882562 Jul  5 20:07 C96_oro_data.tile6.nc
```

sanatcumar commented:
The C96 are the first set of grids generated using steps (1-4 (without the sfc_climogen))
then after the ocean merge C96.mx500 grids are generated.

sanatcumar committed:
fixed a broken path to mosaic

GeorgeGayno-NOAA commented:
I am also seeing extraneous files/links in `/scratch2/NCEPDEV/stmp1/George.Gayno/my_grids_single_step/C96.mx500`

```
lrwxrwxrwx 1 George.Gayno stmp    123 Jul  5 20:07 fort.15 -> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../fix/orog/thirty.second.antarctic.new.bin
lrwxrwxrwx 1 George.Gayno stmp    111 Jul  5 20:07 fort.235 -> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../fix/orog/gmted2010.30sec.int
-rw-r--r-- 1 George.Gayno stmp      0 Jul  5 20:07 fort.25
-rw-r--r-- 1 George.Gayno stmp    251 Jul  5 20:09 input.nml
lrwxrwxrwx 1 George.Gayno stmp    109 Jul  5 20:07 landcover30.fixed -> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../fix/orog/landcover30.fixe
```

`out_dir` is not a working directory; that is `TEMP_DIR`. It is where the final files are stored.

sanatcumar commented:
> I am also seeing extraneous files/links in `/scratch2/NCEPDEV/stmp1/George.Gayno/my_grids_single_step/C96.mx500`
> 
> ```
> lrwxrwxrwx 1 George.Gayno stmp    123 Jul  5 20:07 fort.15 -> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../fix/orog/thirty.second.antarctic.new.bin
> lrwxrwxrwx 1 George.Gayno stmp    111 Jul  5 20:07 fort.235 -> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../fix/orog/gmted2010.30sec.int
> -rw-r--r-- 1 George.Gayno stmp      0 Jul  5 20:07 fort.25
> -rw-r--r-- 1 George.Gayno stmp    251 Jul  5 20:09 input.nml
> lrwxrwxrwx 1 George.Gayno stmp    109 Jul  5 20:07 landcover30.fixed -> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../fix/orog/landcover30.fixe
> ```
> 
> `out_dir` is not a working directory; that is `TEMP_DIR`. It is where the final files are stored.

I would like some advice on this. I found that these files are helpful in knowing what was used to create them and they are created by the original codes as we. I do agree this is perhaps not required and can be removed using a clean up routine ? .

GeorgeGayno-NOAA commented:
> The smaller files are the oro files before ocean merge. They will be different after the merge. Unsure if the sizes should be the same ? Using the default settings: res=96 , ocn=mx500 The C96 are the first set of grids generated using steps (1-4 (without the sfc_climogen)) Then after the ocean merge C96.mx500 grids, oro and corresponding fix are in the C96.mx500 directory. I tried to keep the oro files devoid of the mxYYY for it to be compatible with the chgres procedure for IC generation
> 
> Further, if no ocn is set (ocn="") the code will work and create the Cres directory only but will also include the fix_sfc subdirectory (steps 1-5) just like the original code.

When setting 'ocn', I expected one set of files that include the adjustment from the ocean grid. Outputting multiple sets of files has the potential to be confusing to users.

sanatcumar commented:
> > The smaller files are the oro files before ocean merge. They will be different after the merge. Unsure if the sizes should be the same ? Using the default settings: res=96 , ocn=mx500 The C96 are the first set of grids generated using steps (1-4 (without the sfc_climogen)) Then after the ocean merge C96.mx500 grids, oro and corresponding fix are in the C96.mx500 directory. I tried to keep the oro files devoid of the mxYYY for it to be compatible with the chgres procedure for IC generation
> > Further, if no ocn is set (ocn="") the code will work and create the Cres directory only but will also include the fix_sfc subdirectory (steps 1-5) just like the original code.
> 
> When setting 'ocn', I expected one set of files that include the adjustment from the ocean grid. Outputting multiple sets of files has the potential to be confusing to users.
Do we want to retain the ability to use without the "ocn".?   if yes we want this to be back compatible, then perhaps I can set the steps 1- 5 into the temp directory and then create just one directory.  
else i can insert a clean up tp remove ?? 

What is ideal in your opinion ?

GeorgeGayno-NOAA commented:
> > I am also seeing extraneous files/links in `/scratch2/NCEPDEV/stmp1/George.Gayno/my_grids_single_step/C96.mx500`
> > ```
> > lrwxrwxrwx 1 George.Gayno stmp    123 Jul  5 20:07 fort.15 -> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../fix/orog/thirty.second.antarctic.new.bin
> > lrwxrwxrwx 1 George.Gayno stmp    111 Jul  5 20:07 fort.235 -> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../fix/orog/gmted2010.30sec.int
> > -rw-r--r-- 1 George.Gayno stmp      0 Jul  5 20:07 fort.25
> > -rw-r--r-- 1 George.Gayno stmp    251 Jul  5 20:09 input.nml
> > lrwxrwxrwx 1 George.Gayno stmp    109 Jul  5 20:07 landcover30.fixed -> /scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/driver_scripts/../fix/orog/landcover30.fixe
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > `out_dir` is not a working directory; that is `TEMP_DIR`. It is where the final files are stored.
> 
> I would like some advice on this. I found that these files are helpful in knowing what was used to create them and they are created by the original codes. I do agree this is perhaps not required and can be removed using a clean up routine ? .

I see you added a 'readme' file to the root directory. Perhaps you can add this information to that file.

GeorgeGayno-NOAA commented:
> > > The smaller files are the oro files before ocean merge. They will be different after the merge. Unsure if the sizes should be the same ? Using the default settings: res=96 , ocn=mx500 The C96 are the first set of grids generated using steps (1-4 (without the sfc_climogen)) Then after the ocean merge C96.mx500 grids, oro and corresponding fix are in the C96.mx500 directory. I tried to keep the oro files devoid of the mxYYY for it to be compatible with the chgres procedure for IC generation
> > > Further, if no ocn is set (ocn="") the code will work and create the Cres directory only but will also include the fix_sfc subdirectory (steps 1-5) just like the original code.
> > 
> > 
> > When setting 'ocn', I expected one set of files that include the adjustment from the ocean grid. Outputting multiple sets of files has the potential to be confusing to users.
> > Do we want to retain the ability to use without the "ocn".?   if yes we want this to be back compatible, then perhaps I can set the steps 1- 5 into the temp directory and then create just one directory.
> > else i can insert a clean up tp remove ??
> 
> What is ideal in your opinion ?

For now, I would retain the ability to run without the 'ocn' set. The user should choose his settings, then the scripts should output one set of 'orog' files.

GeorgeGayno-NOAA commented on driver_scripts/log.fv3_grid_driver:
We don't want log files in the repository.

sanatcumar committed:
removed logs

sanatcumar committed:
relocated readme, reorganized output directories, removed logs and other unnecessary temp files & dirs

sanatcumar committed:
added more explanatory comments to code

sanatcumar committed:
Major revision split orog

sanatcumar committed:
Major revision split orog

sanatcumar committed:
Removed logs

sanatcumar committed:
removed some old namelist files

GeorgeGayno-NOAA committed:
Merge branch 'develop' into bugfix/bnu_soil

Fixes #821.

GeorgeGayno-NOAA committed:
Some script updates.

Fixes #821.

sanatcumar committed:
included doxygen(TNX George)+ set ocn default

sanatcumar committed:
updated fv3gfs_driver_grid.sh comments to the correct sequence of calls

sanatcumar committed:
updated fv3gfs_driver_grid.sh comments to the correct sequence of calls

GeorgeGayno-NOAA committed:
Update orion driver script to use additional resources
as required by the new 30s data.

Fixes #821.

GeorgeGayno-NOAA committed:
Update jet grid driver script to request more resources
as required by the new 30s data.

Fixes #821.

GeorgeGayno-NOAA committed:
Minor updates to the wcoss2 driver grid script.

Fixes #821.

GeorgeGayno-NOAA committed:
Merge branch 'develop' into bugfix/bnu_soil

Fixes #821.

GeorgeGayno-NOAA committed:
Update some script comments. Update sfc_climo_gen script to use
the 0.5-degree version of the soil substrate temperature
which will allow the use of more MPI tasks.

Fixes #821.

GeorgeGayno-NOAA committed:
Update comments in the sfc_climo_gen utility script.

Fixes #821.

sanatcumar committed:
clean up before merge

sanatcumar committed:
Update ccpp-physics submodule.

sanatcumar committed:
Update ccpp-physics

sanatcumar committed:
Merge branch 'develop' into single_step

sanatcumar committed:
removed unnecessary files

sanatcumar committed:
fixed a gnu compile issue split the ocean merge into a driver and script

sanatcumar committed:
cleaned up the driver, fix gnu compile, split the ocean merge

GeorgeGayno-NOAA commented on sorc/ocean_merge.fd/merge_lake_ocnmsk.f90:
```diff
@@ -0,0 +1,181 @@
+  open(unit=unit, file='input.nml', iostat=io_status )
+  read(unit,mask_nml, iostat=io_status )
+  close(unit)
+  if (io_status > 0) call handle_err(-1, 'Error reading input.nml')
```
I am getting a warning here. Routine handle_err only takes one argument. The character sting should be removed from the argument list. That string could be printed before the call to handle_err.

sanatcumar replied:
removed that and added a print statement to o/p the error message inside the if statement

GeorgeGayno-NOAA replied:
Upon closer inspection, routine handle_err is for netcdf errors. In this case, the minus 1 return code will be converted to a netcdf error message, which will be misleading to the user. Maybe, you should not call that routine and do something like this instead:

```
if (io_status /= 0) then
  print*,"FATAL ERROR reading namelist input.nml"
  stop 999
endif
```

sanatcumar replied:
Thanks for the insightful tip. Noted. I will make suggested change and test it.  I will also clean up that code in the next commit

sanatcumar committed:
Merge branch 'bugfix/bnu_soil' into single_step

GeorgeGayno-NOAA commented on util/gdas_init/gg/gdas_init/driver.s4.sh:
What are the files in the 'gg' sub-folder? I don't think we want to baseline these.

sanatcumar replied:
Thanks for catching the vestigial folders that had some of my backup scripts I got from you.  Cleaned them all out. currently testing the codes after the chances. Will commit once it runs sucessfully

sanatcumar committed:
Cleaned up old directories and included print messages for debugging in the merge_lake_ocnmsm.f90

GeorgeGayno-NOAA commented:
I have been trying to run other grid options, such as global 'stretched'. Some of the logic will need to change. In script `./ush/fv3gfs_driver_script.sh` for example, I did:

```
-
-
-               out_dir=$out_dir/C$res.mx$ocn
-
-                readme_name=readme.C$res.mx$ocn.txt
-
-
+  if [ -z ${ocn+x} ]; then
+    out_dir=$out_dir/C$res
+    readme_name=readme.C$res.txt
+  else
+    out_dir=$out_dir/C$res.mx$ocn
+    readme_name=readme.C$res.mx$ocn.txt
+  fi
```

and

```
-
-
-   $script_dir/fv3gfs_ocean_merge.sh
+  if [ ! -z ${ocn+x} ]; then
+    $script_dir/fv3gfs_ocean_merge.sh
+  fi

   set +x
-  echo "End uniform orography generation at `date`"
+  echo "End orography generation at `date`"
   set -x
```

sanatcumar committed:
fixed possible misleading errors in merge_lake_ochmsk.f90, fixed logic for readme file namesfor runnig other options

sanatcumar committed:
inlcudes a readme file  for all options

sanatcumar committed:
included logic to bypass when ocn is not set

sanatcumar committed:
Do not use the prev #.

sanatcumar committed:
Now can run other grid type options as usual

sanatcumar committed:
Fixed issues with running other versions

sanatcumar committed:
Updated the comments in the driver scripts

sanatcumar committed:
updated origin and formated coments

sanatcumar committed:
Merge branch 'ufs-community:develop' into single_step

sanatcumar committed:
Merge remote-tracking branch 'upstream/develop' into develop

sanatcumar committed:
Merge branch 'develop' into single_step

sanatcumar committed:
Modified the ocean merge and scriptto pass the binary_lake

sanatcumar committed:
Merge remote-tracking branch 'upstream/develop' into single_step
Merging the develop

sanatcumar committed:
changed the output file names to follow /scratch1/NCEPDEV/global/glopara/fix/orog/20230615/

sanatcumar committed:
Fixed a typo in file name changes

sanatcumar committed:
included the lake source data in the readme

DeniseWorthen commented:
@sanatcumar I never really understood the previous procedure, but I know that it used the mapped ocean mask files generated by the cpld_gridgen utility. Those files give the MOM6 land mask (ocean fraction) on the ATM tiles and have names containing the ATM resolution (eg C96), the ocean resolution (eg mx100) and the tile number. Does this procedure still use those files, so that it is dependent on cpld_gridgen generating them?

sanatcumar commented:
> @sanatcumar I never really understood the previous procedure, but I know that it used the mapped ocean mask files generated by the cpld_gridgen utility. Those files give the MOM6 land mask (ocean fraction) on the ATM tiles and have names containing the ATM resolution (eg C96), the ocean resolution (eg mx100) and the tile number. Does this procedure still use those files, so that it is dependent on cpld_gridgen generating them?

I used the cpld gridgen to pregenerate the "025","050","100" & 500  ocn . and this single step uses these pre generated files and does not make them on the fly.


Hope this clarifies. 
Sanath

DeniseWorthen commented:
OK, great. Then I will close #793, which I had thought would be necessary. But it appears to have been superceded by your work. Thanks.

sanatcumar committed:
fixed file names to run gsl suite

sanatcumar committed:
Merge remote-tracking branch 'upstream/develop' into single_step

sanatcumar committed:
Included logic to work with different filenames

sanatcumar committed:
Fixed issues with nck commands with the new libaries from last PR

sanatcumar committed:
Made filenames consistent

sanatcumar committed:
Included creation date to readme

sanatcumar committed:
Includes Mike's fix for coastal ORO issue
 Please enter the commit message for your changes. Lines starting

sanatcumar committed:
Includes Mike's coastal oro fix

sanatcumar committed:
Merge remote-tracking branch 'upstream/develop' into single_step

sanatcumar committed:
tweaks to get the grid_gen test working

sanatcumar committed:
inlcuding lake source data for all gtypes

sanatcumar committed:
Merge remote-tracking branch 'upstream/develop' into single_step

sanatcumar committed:
Tweaks to get gridgen tests working. Includes checking for uniform and if ocn is set to enable the ocean merge

GeorgeGayno-NOAA commented:
I was getting Doxygen warnings. Can you check in the following files from `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath`

- `sorc/ocean_merge.fd/merge_lake_ocnmsk.f90`
- `sorc/orog_mask_tools.fd/orog.fd/mtnlm7_oclsm.F`
- `sorc/orog_mask_tools.fd/orog.fd/netcdf_io.F90`

sanatcumar commented:
> I was getting Doxygen warnings. Can you check in the following files from `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath`
> 
> * `sorc/ocean_merge.fd/merge_lake_ocnmsk.f90`
> * `sorc/orog_mask_tools.fd/orog.fd/mtnlm7_oclsm.F`
> * `sorc/orog_mask_tools.fd/orog.fd/netcdf_io.F90`

 I am looking into it and I see a similar warning for 
[ 98%] Generating API Documentation with Doxygen
Warning: ignoring unsupported tag `ALLOW_UNICODE_NAMES    =' at line 79, file /scratch2/NCEPDEV/land/Sanath.Kumar/single_step/UFS_UTILS/build/sorc/weight_gen.fd/docs/Doxyfile

Related or related issue ?

sanatcumar committed:
includes Doxygen edits

GeorgeGayno-NOAA commented on driver_scripts/driver_grid.orion.sh:
```diff
@@ -83,19 +83,31 @@ export vegsoilt_frac='.false.'  # When true, outputs percent of each
                                # outputs the dominant category. A
                                # Fortran logical, so include the dots.
 
```
These lines - <<< HEAD and >>> develop - are the result of a merge conflict. Please clean up.

GeorgeGayno-NOAA replied:
Other driver scripts have the same issue.

sanatcumar replied:
Done

GeorgeGayno-NOAA commented on driver_scripts/driver_grid.hera.sh:
```diff
@@ -115,6 +115,7 @@ if [ $gtype = uniform ]; then
   export add_lake=true         # Add lake frac and depth to orography data.
   export lake_cutoff=0.50      # return 0 if lake_frac <  lake_cutoff & add_lake=T
   export binary_lake=1         # return 1 if lake_frac >= lake_cutoff & add_lake=T
+  export ocn=${ocn:-"100"}     # use one of  "025", "050", "100", "500". Cannot be empty
```
This line should be added to the scripts for the other machines.

sanatcumar replied:
Noted. @GeorgeGayno-NOAA , Please advice on handling this ocean merge files in other machines and in Hera. 
Once a path is decided we can update the paths for each machine specific script 
for Hera its here and for Orion I created a copy in my work space 
/scratch1/NCEPDEV/stmp4/Sanath.Kumar/ocean_mask/CPLD_GRIDGEN/

GeorgeGayno-NOAA replied:
Those files are 'fixed' correct? If so, they should be placed somewhere in the official 'fixed' directories.

sanatcumar replied:
Yes they are pregeneated and so "fixed". They can be regenerated and I have included a README in that directrory. 
If it needs to placed in an official location how do i Go about it ? I will  clean it up as it does have some unnecessary files too.

GeorgeGayno-NOAA replied:
When those ocean merge files are missing, the program fails, but the script continues. I added some error handling logic to two scripts so execution ceases. 

```
--- a/ush/fv3gfs_ocean_merge.sh
+++ b/ush/fv3gfs_ocean_merge.sh
@@ -15,12 +15,17 @@
     /
 EOF

+    echo Run ocean_merge program.

     time  ${exec_dir}/./ocean_merge
+    rc=$?

-    set -x
-    echo run orog 2nd time
-    set +x
+    if [[ $rc -ne 0 ]] ; then
+     echo FATAL ERROR running ocean_merge.
+     exit $rc
+    fi
+
+    echo Run orog 2nd time

     for tnum in '1' '2' '3' '4' '5' '6'
     do
```
 
and 

```
--- a/ush/fv3gfs_driver_grid.sh
+++ b/ush/fv3gfs_driver_grid.sh
@@ -226,12 +226,15 @@ if [ $gtype = uniform ] || [ $gtype = stretch ] || [ $gtype = nest ];  then
     fi
   fi

-
-       if [ $gtype = uniform ]; then
-               if declare -p ocn &>/dev/null;then
-                        $script_dir/fv3gfs_ocean_merge.sh
-               fi
-       fi
+  if [ $gtype = uniform ]; then
+    if declare -p ocn &>/dev/null;then
+      $script_dir/fv3gfs_ocean_merge.sh
+      err=$?
+      if [ $err != 0 ]; then
+        exit $err
+      fi
+    fi
+  fi

   set +x
```

You can find the updated script on Orion - /work/noaa/da/ggayno/save/ufs_utils.git/UFS_UTILS.sanath/ush

GeorgeGayno-NOAA replied:
> Yes they are pregeneated and so "fixed". They can be regenerated and I have included a README in that directrory. If it needs to placed in an official location how do i Go about it ? I will clean it up as it does have some unnecessary files too.

Create a clean directory (no extraneous files) and we will ask @KateFriedman-NOAA for assistance.

GeorgeGayno-NOAA replied:
> > Yes they are pregeneated and so "fixed". They can be regenerated and I have included a README in that directrory. If it needs to placed in an official location how do i Go about it ? I will clean it up as it does have some unnecessary files too.
> 
> Create a clean directory (no extraneous files) and we will ask @KateFriedman-NOAA for assistance.

We can open an issue here: https://github.com/NOAA-EMC/global-workflow/issues/new/choose

sanatcumar replied:
Will make a clean  directory, test it out and then will start the process

sanatcumar committed:
Included newer variables in driver scripts for other machines

sanatcumar committed:
Updated path for ocean merge files on orion

sanatcumar committed:
Includes logic by George to stop on error in the ocean merge routines

GeorgeGayno-NOAA commented on util/sfc_climo_gen/sfc_gen.sh:
```diff
@@ -39,7 +39,11 @@
 #
 # veg_type_src - Input vegetation type data. Choices are:
 #                  For viirs-based vegetation type data, set to:
```
Fix the merge conflicts.

GeorgeGayno-NOAA commented on util/gdas_init/run_v16.chgres.sh:
```diff
@@ -12,10 +12,18 @@ set -x
 
 MEMBER=$1
 
+
```
Don't forget to remove all references to your personal directories.

GeorgeGayno-NOAA replied:
Not just this routine, but others as well.

GeorgeGayno-NOAA commented on ush/fv3gfs_make_orog.sh:
```diff
@@ -99,6 +99,14 @@ cp $executable .
+  echo ".true." >> INPS
+fi 
+echo "none" >> INPS
+#echo '"'/scratch2/NCEPDEV/stmp1/George.Gayno/fv3_grid.uniform.ocean.merge/ocean_merged/C96.mx100/C96.mx100.tile${tile}.nc'"' >> INPS
```
Commented out code should be removed.

sanatcumar replied:
Done

GeorgeGayno-NOAA commented on docs/source/ufs_utils.rst:
```diff
@@ -544,14 +544,22 @@ The surface climatological data is located here `./fix/fix_sfc_climo <https://no
              * N Hemis 30 sec - soil_type.statsgo.nh.30s.nc
              * Global 30 sec - soil_type.statsgo.30s.nc
       * Categorical BNU soil type
```
Remove merge conflicts.

sanatcumar replied:
I removed a conflict in sfc climo.  I see these merge conflicts in many files. Removing them now

sanatcumar committed:
Removed personal links except "ocean merge" and merged conflicts

GeorgeGayno-NOAA commented on util/gdas_init/config:
```diff
@@ -51,15 +51,15 @@
 #
 #-----------------------------------------------------------
 
```
Don't forget to remove references to your personal directories.

GeorgeGayno-NOAA commented on util/sfc_climo_gen/run.hera.sh:
```diff
@@ -17,14 +15,110 @@
 export APRUN_SFC="srun"
 
 export BASE_DIR=$SLURM_SUBMIT_DIR/../..
 
```
Why was this file updated? Just for your testing?

sanatcumar replied:
yes. I will revert

GeorgeGayno-NOAA replied:
Don't forget to revert.

GeorgeGayno-NOAA commented on ush/sfc_climo_gen.sh:
```diff
@@ -49,23 +49,32 @@ VEG_TYPE_FILE=${VEG_TYPE_FILE:-${input_sfc_climo_dir}/vegetation_type.${veg_type
 
 rm -fr $WORK_DIR
 mkdir -p $WORK_DIR
 cd $WORK_DIR
```
Commented out code.

sanatcumar replied:
removed all commented codes

sanatcumar commented:
> To get the regression tests to work, I had to update these scripts.
> 
> * c96.viirs.bnu.sh
> * esg.regional.pct.cat.sh
> * esg.regional.sh
> * gfdl.regional.sh
> * regional.gsl.gwd.sh
> 
> Also, since the default for uniform grids will now include the ocean merge step, I want to update this regression test to run that step:
> 
> * c96.uniform.sh
> 
> Please find all updated scripts on Orion here: `/work/noaa/da/ggayno/save/ufs_utils.git/UFS_UTILS.sanath/reg_tests/grid_gen`:

 Hi George, although the "ocn " is defined in the driver, it is not declared as a default in the /ush/fv3gfs_driver*.sh. This way the grid gen test 1 will run and produce results that fails the original only because of the different orog . 

I agree adding a test is the way forward. I will copy over these files and merge them in this. Thanks.

sanatcumar committed:
Changes to test the new location for ocean masks

GeorgeGayno-NOAA commented on driver_scripts/driver_grid.hera.sh:
```diff
@@ -150,14 +151,20 @@ fi
 export TEMP_DIR=/scratch2/NCEPDEV/stmp1/$LOGNAME/fv3_grid.$gtype
-export out_dir=/scratch2/NCEPDEV/stmp1/$LOGNAME/my_grids
+export out_dir=/scratch2/NCEPDEV/stmp1/$LOGNAME/my_grids_ocean_fix/
+#export ocean_mask_dir=/scratch1/NCEPDEV/stmp4/Sanath.Kumar/ocean_mask/CPLD_GRIDGEN/
```
I would set ocean_mask_dir in fv3gfs_driver_grid.sh using the $home_dir variable to set the path. Then, you don't need to declare it in every machine's driver script.

sanatcumar replied:
HI @GeorgeGayno-NOAA ,
 I had to use dirname in /ush/fv3gfs_ocean_merge.sh as the fortran merge code could not handle the "../" relative paths.
     ocean_mask_dir="$(dirname $(dirname $home_dir))/fix/orog/C${res}/ocean_mask/${ocn}/""
It works now. But is there a better solution

GeorgeGayno-NOAA replied:
Try this:

Update `fv3gfs_ocean_merge.sh` as follows:

```
     cat << EOF > input.nml
      &mask_nml
-     ocean_mask_dir="$(dirname $(dirname $home_dir))/fix/orog/C${res}/ocean_mask/${ocn}/"
+     ocean_mask_dir="${home_dir}/fix/orog/C${res}/ocean_mask/${ocn}/"
      ocnres="mx${ocn}"
      lake_mask_dir="${TEMP_DIR}/C${res}/orog/"
      atmres="C${res}"
@@ -22,6 +23,8 @@ EOF
```

and `merge_lake_ocnmsk.f90` as follows

```
   integer :: binary_lake

-  character(len=120) :: flnm
+  character(len=250) :: flnm
   integer :: ncid,ndims,nvars,natts,lat,lon,v1id,v2id,v3id,v4id,start(2),count(2),i,j,latid,lonid,ncid4, dims(2),tile,nodp_pt
```

sanatcumar replied:
This worked. Thanks @GeorgeGayno-NOAA

sanatcumar committed:
testing with new fix locations

sanatcumar committed:
edited driver scripts for each machine

sanatcumar committed:
Tweaks to make it run with the new soft linked orog

sanatcumar committed:
Cleaned up unused variables

sanatcumar committed:
Cleaning up the unnecessary changes to use the fix files

sanatcumar committed:
Tweaks recomended by George to read long filenames

GeorgeGayno-NOAA commented on ush/fv3gfs_driver_grid.sh:
```diff
@@ -521,4 +566,89 @@ if [ $gtype = nest ]; then
+
+cat <<EOF > $readme_name
+The following # was used
+https://github.com/sanatcumar/UFS_UTILS/tree/single_step
```
References to your github branch won't work after the merge.

GeorgeGayno-NOAA replied:
I just thought of something. The logic in `./reg_tests/get_hash.sh` is used to get the github commit hash. Could the hash be added as part of the readme file?

sanatcumar replied:
That worked great!

GeorgeGayno-NOAA commented:
How are you pointing to the new fixed directory without updating `./fix/link_fixdirs.sh`

GeorgeGayno-NOAA commented on util/gdas_init/config:
You added references to your personal directories. Please remove. This file should not change under this PR.

GeorgeGayno-NOAA commented on ush/sfc_climo_gen.sh:
```diff
@@ -49,23 +49,32 @@ VEG_TYPE_FILE=${VEG_TYPE_FILE:-${input_sfc_climo_dir}/vegetation_type.${veg_type
+	else
+		the_orog_files='"C'${res}'_oro_data.tile1.nc","C'${res}'_oro_data.tile2.nc","C'${res}'_oro_data.tile3.nc","C'${res}'_oro_data.tile4.nc","C'${res}'_oro_data.tile5.nc","C'${res}'_oro_data.tile6.nc"'
+	fi
 fi
```
Commented out code should be removed.

sanatcumar replied:
Noted. I thought I had removed all personal references. I will check and fix

sanatcumar committed:
Reverted non relevant edits for this PR and removed a personal link

sanatcumar committed:
Removed personal link in readme

GeorgeGayno-NOAA commented:
Since the name of the `fix_sfc` subdirectory was changed to `sfc`, I had to make a minor change the `grid_gen` regression test scripts. Find them in `/scratch1/NCEPDEV/da/George.Gayno/ufs_utils.git/UFS_UTILS.sanath/reg_tests/grid_gen`

- c96.uniform.sh
- c96.viirs.bnu.sh
- esg.regional.pct.cat.sh
- esg.regional.sh
- gfdl.regional.sh
- regional.gsl.gwd.sh

sanatcumar committed:
Includes edits by george to grid_gen tests to accomodate the change from fix_sfc to sfc

sanatcumar committed:
Tweaks to inlcude the hash automatically in the readme file. Nice idea George

sanatcumar committed:
last edits to sfc_gen

GeorgeGayno-NOAA commented:
The regression tests were run on all Tier 1 machines using d0d1e8f. Only the `grid_gen` tests failed. This result was expected. The 'grid' and 'sfc climo' files were identical to the baseline. But the 'oro' showed differences. These differences were the result of updates to the `orog` code. (@sanatcumar can explain in more detail). The tile1 differences from the c96.uniform tests were:

```
Variable  Group Count      Sum  AbsSum         Min        Max   Range     Mean  StdDev
orog_raw  /       353  58512.3 58512.3 -0.00636026    979.795 979.801  165.757 167.988
orog_filt /      1008  58437.8 61182.7    -93.6286    978.194 1071.82   57.974 122.876
elvmax    /       352 -58088.7 58088.7    -979.795 0.00636101 979.801 -165.025 168.336
```

Other `grid_gen` tests showed similar differences.

sanatcumar commented:
> The regression tests were run on all Tier 1 machines using [d0d1e8f](https://github.com/ufs-community/UFS_UTILS/commit/d0d1e8f47ff1469c249c96d091bf51771e64c29e). Only the `grid_gen` tests failed. This result was expected. The 'grid' and 'sfc climo' files were identical to the baseline. But the 'oro' showed differences. These differences were the result of updates to the `orog` code. (@sanatcumar can explain in more detail). The tile1 differences from the c96.uniform tests were:
> 
> ```
> Variable  Group Count      Sum  AbsSum         Min        Max   Range     Mean  StdDev
> orog_raw  /       353  58512.3 58512.3 -0.00636026    979.795 979.801  165.757 167.988
> orog_filt /      1008  58437.8 61182.7    -93.6286    978.194 1071.82   57.974 122.876
> elvmax    /       352 -58088.7 58088.7    -979.795 0.00636101 979.801 -165.025 168.336
> ```
> 
> Other `grid_gen` tests showed similar differences.

Orography is set to zero when lake fraction and slmask is zero in the original codes. This causes some fractional land points near oceans to have zero orography with the inclusion of newer codes to generate atmosphere and ocean coupled grids. To avoid this error new logic was added to sorc/orog_mask_tools.fd/orog.fd/mtnlm7_oclsm.F to set orography to  zero only when both lake and land frac is zero. Small differences are expected. @barlage please chime in if I have missed anything.

GeorgeGayno-NOAA commented:
@sanatcumar - I have no more comments. Unless you object, I will merge tomorrow.

sanatcumar commented:
> @sanatcumar - I have no more comments. Unless you object, I will merge tomorrow.

I do not have any further comments/ suggestions. Thanks George
